src/Compilers/CSharp/Portable/Binder/Binder_Operators.cs

```diff
@@ -2233,7 +2233,7 @@ private BoundExpression BindUnaryOperatorCore(CSharpSyntaxNode node, string oper
                 return new BoundUnaryOperator(node, kind, operand, ConstantValue.NotAvailable,
                     methodOpt: null,
                     resultKind: LookupResultKind.Empty,
-                    type: GetSpecialType(SpecialType.System_Object, diagnostics, node),
+                    type: operand.Type ?? GetSpecialType(SpecialType.System_Object, diagnostics, node),
```

Comment:
We should be using an error type here. That is exactly what error types are for.